cluster/assemblestate/selector.go

```diff
@@ -0,0 +1,462 @@
+	for eid, edge := range p.edges.Values() {
+		fromRDT := p.rdts.Value(edge.from)
+		toRDT := p.rdts.Value(edge.to)
+
+		if p.identified(fromRDT) && p.identified(toRDT) {
+			p.verifiedEdges.Add(edgeID(eid))
+		}
+	}
+}
+
+// Select n selects routes that should be published to the specified peer.
+//
+// We prioritize routes that this local node has witnessed and published fewer
+// times.
+func (p *PrioritySelector) Select(to DeviceToken, n int) (routes Routes, ack func(), ok bool) {
+	selected, exists := p.rdts.IndexOf(to)
+	if !exists {
+		return Routes{}, nil, false
+	}
+
+	if to == p.self {
+		return Routes{}, nil, false
+	}
+
+	peerKnown := p.knownByPeers[selected]
+	unknown := p.verifiedEdges.Diff(peerKnown)
+
+	// max possible needed size is all unknown routes + the route from this
+	// local node to the destination peer
+	edgesToSend := make([]edgeID, 0, unknown.Count()+1)
+
+	// only consider routes that we don't think that this peer knows about
+	unknown.Range(func(eid edgeID) bool {
+		edgesToSend = append(edgesToSend, eid)
+		return true
+	})
+
+	// prioritize routes that we think fewer peers know about. this takes into
+	// account copies of routes we've seen from our peers and copies that we've
+	// sent to our peers
+	sort.Slice(edgesToSend, func(i, j int) bool { return p.edgeSources[edgesToSend[i]] < p.edgeSources[edgesToSend[j]] })
+
+	// discard any edges causing us to exceed the given threshold. thus, we pick
+	// the n least frequently seen routes.
+	edgesToSend = edgesToSend[:min(len(edgesToSend), n)]
+
+	// if we have an authoritative route for this peer, make sure to include it
+	if eid, ok := p.authoritative[to]; ok && !peerKnown.Contains(eid) {
+		edgesToSend = append(edgesToSend, eid)
+	}
+
+	if len(edgesToSend) == 0 {
+		return Routes{}, nil, false
+	}
+
+	routes = p.edgesToRoutes(edgesToSend)
+
+	// create ack function that updates source counts when called
+	ack = func() {
+		for _, eid := range edgesToSend {
+			// the peer might know about the route already since selection time.
+			// if that has happened, then we don't want to double count that
+			// peer as a source.
+			if !peerKnown.Contains(eid) {
+				peerKnown.Add(eid)
+				p.edgeSources[eid]++
+			}
+		}
+	}
+
+	return routes, ack, true
+}
+
+// Routes returns routes that we've seen for which both peers in the route
+// identified.
+func (p *PrioritySelector) Routes() Routes {
+	eids := p.verifiedEdges.All()
+
+	devs := make([]string, 0, len(p.rdts.Values()))
+	devIndexes := make(map[string]int, cap(devs))
+
+	addrs := make([]string, 0, len(p.addresses.Values()))
+	addrIndexes := make(map[string]int, cap(addrs))
+
+	// here we build a slice of devices and address that are used in the routes.
+	// we want the slices to be sorted for output consistency, so we also build
+	// a mapping of values to indexes in these slices. this first pass just
+	// fills the maps with placeholders (-1) until we can actually assign those
+	// values after sorting.
```

Comment:
maybe this way is clearer: "... (-1) to track seen values, later we actually assign ..." ?